text/3660-crates-io-crate-deletions.md

```diff
@@ -0,0 +1,123 @@
+
+# Prior art
+
+## npm
+
+The main inspiration for this proposal comes from the npm registry, which has a similar policy for deleting packages and versions:
+
+- https://docs.npmjs.com/policies/unpublish
+- https://docs.npmjs.com/unpublishing-packages-from-the-registry
+
+The npm registry started with a more permissive policy, but had to tighten it over time. 
+It started out with a policy that allowed package owners to delete their packages at any time, but this led to a number of issues, [such as packages being deleted that were depended on by other packages](https://en.wikipedia.org/wiki/Npm_left-pad_incident).
+Their policy was later changed to require that packages can only be deleted within 72 hours of being published, and then [changed again in January 2020](https://blog.npmjs.org/post/190553543620/changes-to-npmunpublish-policy-january-2020) to allow deletions outside the 72-hour window under certain conditions.
+
+
+## PyPI
+
+The Python Package Index (PyPI) still allows package owners to delete their packages (or a subset of released files) at any time.
+A member of the PyPI team has proposed to [stop allowing deleting things from PyPI](https://discuss.python.org/t/stop-allowing-deleting-things-from-pypi/17227) due to the same issues that the npm registry faced. The most current proposed ruleset can be found [here](https://discuss.python.org/t/stop-allowing-deleting-things-from-pypi/17227/71).
+
+Their proposal is also inspired by the npm registry policy, but notably does not include a reverse dependency restriction. It seems that PyPI might not currently be tracking dependencies between packages, which would make it harder for them to implement such a restriction.
+
+## Others
+
+<https://discuss.python.org/t/stop-allowing-deleting-things-from-pypi/17227/59> contains a list of other package registries and their deletion policies.
+
+
+# Unresolved questions
+
+## Should names of deleted crates be blocked so that they can't be re-used?
+
+The reason for this would be to prevent someone else from re-publishing a crate with the same name, which could lead to potential security issues.
+Due to the restrictions on the number of downloads and reverse dependencies, this seems like a low risk though.
+The advantage of allowing others to re-use such names is that it allows name-squatted/placeholder crates to be released back to the community without the crates.io team having to manually intervene.
+
+The npm registry blocks re-use of deleted package names for 24 hours.
+
+
+## Should deleted versions be blocked from being re-uploaded?
+
+Since version deletions would also be possible for widely used crates, it might make sense to block re-uploads of deleted versions to prevent security issues.
+However, this would make it impossible to fix a mistakenly published new major version, for example.
```

Comment:
I'm not sure what this impossibility is referencing? If you publish a `2.0.0`, realize you left an API token or something in it and delete it, then you can still publish a `2.0.1` as a new major version. Blocking a single version number indefinitely seems fine to me as there's ~infinite mostly equivalent numbers to use instead (the one case where there isn't is if you were publishing with gaps in your patch numbers and going back to fill in a missing patch number, but I've never seen that done with semver).

Reply:
I agree. And there's a lot of advantage in saying "this old version might stop existing, but it'll never show up again with a different hash". (Also keep in mind that people might have cached local builds with the old version.)

Reply:
one place where i could see that running into problems is where you're trying to track exact versions of some 3rd-party library, where if they publish a particular version, you'd want to publish the exact same version number even if you mistakenly published something with that version number before

Reply:
@Nemo157 I **personally** agree, but I've seen enough cases of "the marketing team" pushing for "the big 2.0.0 release" and if that version number has already been burned it becomes a problem. But that's why it's in the "Unresolved questions" section, I'm open to input on what the community prefers. I tend to follow what other successful projects/communities/registries are doing unless there are good reasons against it, and since npm blocks deleted version numbers I would personally vote for that option too.